src/read.c

```diff
@@ -5422,18 +5417,13 @@ avifResult avifDecoderReset(avifDecoder * decoder)
                 // Read the gain map's metadata.
                 avifROData tmapData;
                 AVIF_CHECKRES(avifDecoderItemRead(toneMappedImageItem, decoder->io, &tmapData, 0, 0, data->diag));
-                if (decoder->image->gainMap == NULL) {
-                    decoder->image->gainMap = avifGainMapCreate();
-                    AVIF_CHECKERR(decoder->image->gainMap, AVIF_RESULT_OUT_OF_MEMORY);
-                }
+                AVIF_ASSERT_OR_RETURN(decoder->image->gainMap != NULL);
                 const avifResult tmapParsingRes =
                     avifParseToneMappedImageBox(&decoder->image->gainMap->metadata, tmapData.data, tmapData.size, data->diag);
-                if (tmapParsingRes != AVIF_RESULT_OK) {
-                    avifGainMapDestroy(decoder->image->gainMap);
-                    decoder->image->gainMap = NULL;
-                }
                 if (tmapParsingRes == AVIF_RESULT_NOT_IMPLEMENTED) {
                     // Unsupported gain map version. Simply ignore the gain map.
+                    avifGainMapDestroy(decoder->image->gainMap);
+                    decoder->image->gainMap = NULL;
```

Comment:
If we return from the `AVIF_CHECKRES` at line 5428, is `decoder->image->gainMap` guaranteed to be NULL?

Reply:
If AVIF_CHECKRES returns an error, decoder->image->gainMap might be set even if gainMapPresent is false (inconsistent state). But If the function returns an error, it doesn't have to guarantee anything about the output, does it?